pkg/resources/grant_application_role.go

```diff
@@ -0,0 +1,199 @@
+					return nil, fmt.Errorf("invalid ID specified: %v, expected <name>|<object_type>|<target_identifier>", d.Id())
+				}
+				if err := d.Set("name", parts[0]); err != nil {
+					return nil, err
+				}
+				switch parts[1] {
+				case "ROLE":
+					if err := d.Set("parent_role_name", strings.Trim(parts[2], "\"")); err != nil {
+						return nil, err
+					}
+				case "APPLICATION":
+					if err := d.Set("application_name", strings.Trim(parts[2], "\"")); err != nil {
+						return nil, err
+					}
+				default:
+					return nil, fmt.Errorf("invalid object type specified: %v, expected ROLE, APPLICATION", parts[1])
+				}
+
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+	}
+}
+
+func CreateGrantApplicationRole(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*provider.Context).Client
+	ctx := context.Background()
+	name := d.Get("name").(string)
+	applicationRoleIdentifier := sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(name)
+	// format of snowflakeResourceID is <application_role_identifier>|<object type>|<parent_role_name>
+	var snowflakeResourceID string
+	if parentRoleName, ok := d.GetOk("parent_role_name"); ok && parentRoleName.(string) != "" {
+		parentRoleIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(parentRoleName.(string))
+		snowflakeResourceID = helpers.EncodeSnowflakeID(applicationRoleIdentifier.FullyQualifiedName(), sdk.ObjectTypeRole.String(), parentRoleIdentifier.FullyQualifiedName())
+		req := sdk.NewGrantApplicationRoleRequest(applicationRoleIdentifier).WithTo(*sdk.NewKindOfRoleRequest().WithRoleName(&parentRoleIdentifier))
+		if err := client.ApplicationRoles.Grant(ctx, req); err != nil {
+			return err
+		}
+	} else if applicationName, ok := d.GetOk("application_name"); ok && applicationName.(string) != "" {
+		applicationIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(applicationName.(string))
+		snowflakeResourceID = helpers.EncodeSnowflakeID(applicationRoleIdentifier.FullyQualifiedName(), sdk.ObjectTypeApplication.String(), applicationIdentifier.FullyQualifiedName())
+		req := sdk.NewGrantApplicationRoleRequest(applicationRoleIdentifier).WithTo(*sdk.NewKindOfRoleRequest().WithApplicationName(&applicationIdentifier))
+		if err := client.ApplicationRoles.Grant(ctx, req); err != nil {
+			return err
+		}
+	}
+	d.SetId(snowflakeResourceID)
+	return ReadGrantApplicationRole(d, meta)
+}
+
+func ReadGrantApplicationRole(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*provider.Context).Client
+	parts := strings.Split(d.Id(), helpers.IDDelimiter)
+	applicationRoleName := parts[0]
+	applicationRoleIdentifier := sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(applicationRoleName)
+	objectType := parts[1]
+	targetIdentifier := parts[2]
+	ctx := context.Background()
+	grants, err := client.Grants.Show(ctx, &sdk.ShowGrantOptions{
+		Of: &sdk.ShowGrantsOf{
+			ApplicationRole: applicationRoleIdentifier,
+		},
+	})
+	if err != nil {
```

Comment:
We recently did a change to detect missing dependent resources. Right now with SHOW GRANTS you can only detect if the application role is missing, but can't really tell if the parent_role or application is missing. Could you please adjust it to the recent change where in Read we check the existence of other objects (parent_role or application in this case). You can check the code below (it's from the snowflake_grant_privileges_to_account_role):
```go
	if _, err := client.Roles.ShowByID(ctx, sdk.NewShowByIdRoleRequest(id.RoleName)); err != nil && err.Error() == "object does not exist" {
		d.SetId("")
		return diag.Diagnostics{
			diag.Diagnostic{
				Severity: diag.Warning,
				Summary:  "Failed to retrieve account role. Marking the resource as removed.",
				Detail:   fmt.Sprintf("Id: %s", d.Id()),
			},
		}
	}

	logging.DebugLogger.Printf("[DEBUG] About to show grants")
	grants, err := client.Grants.Show(ctx, opts)
	if err != nil {
		if errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
			d.SetId("")
			return diag.Diagnostics{
				diag.Diagnostic{
					Severity: diag.Warning,
					Summary:  "Failed to retrieve grants. Target object not found. Marking the resource as removed.",
					Detail:   fmt.Sprintf("Id: %s", d.Id()),
				},
			}
		}
		return diag.Diagnostics{
			diag.Diagnostic{
				Severity: diag.Error,
				Summary:  "Failed to retrieve grants",
				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err),
			},
		}
	}
```
Also, see that we brought back the error comparison in the show grants error handling to SetId("") only if the sdk.ErrObjectNotExistOrAuthorized is returned.